### PR TITLE
[alphafold] Transpose support large tensors where there numel is bigger than INT32_MAX

### PR DESCRIPTION
**Pull request by FeixLiu**
### PR types
Others

### PR changes
Others

### Describe
Transpose support large tensors where there numel is bigger than INT32::MAX

All branches can be tested with this script

```python
import paddle
import numpy as np

# case 2, large tile
print('testing case 2, large tile')
x = paddle.rand((5800 * 5800, 64))
y = paddle.transpose(x, perm=[1, 0])
xy = paddle.transpose(y, perm=[1, 0])
assert np.array_equal(x.numpy(), xy.numpy())

del x
del y
del xy


# case 2, narrow tile
print('testing case 2, narrow tile')
x = paddle.rand((5800 * 5800 * 8, 8))
y = paddle.transpose(x, perm=[1, 0])
xy = paddle.transpose(y, perm=[1, 0])
assert np.array_equal(x.numpy(), xy.numpy())

del x
del y
del xy


# case 32
print('testing case 32')
x = paddle.rand((5800, 5800, 64))
y = paddle.transpose(x, perm=[2, 1, 0])
xy = paddle.transpose(y, perm=[2, 1, 0])
assert np.array_equal(x.numpy(), xy.numpy())

del x
del y
del xy


# case 31, large tile
print('testing case 31, large tile')
x = paddle.rand((5800, 5800, 64))
y = paddle.transpose(x, perm=[0, 2, 1])
xy = paddle.transpose(y, perm=[0, 2, 1])
assert np.array_equal(x.numpy(), xy.numpy())

del x
del y
del xy


# case 31, narrow tile
print('testing case 31, narrow tile')
x = paddle.rand((4500 * 5000, 1, 96))
y = paddle.transpose(x, perm=[0, 2, 1])
xy = paddle.transpose(y, perm=[0, 2, 1])
assert np.array_equal(x.numpy(), xy.numpy())

del x
del y
del xy


# case 31, no tile
print('testing case 31, no tile')
x = paddle.rand((5800 * 5800, 8, 8))
y = paddle.transpose(x, perm=[0, 2, 1])
xy = paddle.transpose(y, perm=[0, 2, 1])
assert np.array_equal(x.numpy(), xy.numpy())

del x
del y
del xy


# geeral case, fall into egine
print('testing general egine case')
x = paddle.rand((500, 9, 500, 10, 96))
y = paddle.transpose(x, perm=[4, 0, 2, 1, 3])
xy = paddle.transpose(y, perm=[1, 3, 2, 4, 0])
assert np.array_equal(x.numpy(), xy.numpy())

del x
del y
del xy
```

### CODE REVIEW

File: paddle/fluid/framework/gpu_utils.h [hunk 1]
```diff
@@ -82,22 +82,24 @@ struct Index3 : DeviceArray<int, 3, 0> {
...
+template <typename IDX_T = int>
 EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE Index3
-ConvertTensorIndex(int index, const Dim3& dims) {
+ConvertTensorIndex(IDX_T index, const Dim3& dims) {
   Index3 tensor_index;
   for (int i = 2; i >= 0; i--) {
-    int new_index = index / dims[i];
-    tensor_index[i] = index - dims[i] * new_index;
+    IDX_T new_index = index / dims[i];
+    tensor_index[i] = static_cast<int>(index - dims[i] * new_index);
```
Comment (sneaxiy) [inline paddle/fluid/framework/gpu_utils.h:102]: Why cast to `int`?

Comment (FeixLiu) [reply paddle/fluid/framework/gpu_utils.h:102]: The `tensor_index` is an instance of `Index3`, which is int actually. Just avoid some inexplicit casts.

Comment (sneaxiy) [reply paddle/fluid/framework/gpu_utils.h:102]: Seems like it should be fixed further (not in this PR)... This PR can be merged right now.

**Review by sneaxiy — APPROVED** [review]

File: paddle/fluid/framework/gpu_utils.h (hunk 1 above)
Comment (FeixLiu) [reply paddle/fluid/framework/gpu_utils.h:102]: Just remove this cast? Let int64 be implicit cast to int32 when IDX_T is int64?

Comment (sneaxiy) [reply paddle/fluid/framework/gpu_utils.h:102]: No. I mean `Index3` might be `Index3<IDX_T>`.

Comment (FeixLiu) [reply paddle/fluid/framework/gpu_utils.h:102]: Probably in the future when encounter problems 😂, for current usage, use int for `Index3` is enough.

Comment (sneaxiy) [reply paddle/fluid/framework/gpu_utils.h:102]: Agree so far.

File: paddle/fluid/framework/gpu_utils.h [hunk 2]
```diff
@@ -82,22 +82,24 @@ struct Index3 : DeviceArray<int, 3, 0> {
 };
 
 // Flat index with real dimension
-EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE int FlatTensorIndex(const Index3& index,
-                                                          const Dim3& dims) {
-  int flat_index = index[0];
+template <typename IDX_T = int>
```
Comment (Xreki) [inline paddle/fluid/framework/gpu_utils.h:85]: 建议`IDX_T` -> `IndexT`或`IndexType`

File: paddle/fluid/operators/transpose_op.cu.h [hunk 3]
```diff
@@ -1159,7 +1181,13 @@ void TransposeGPUKernelDriver(const phi::GPUContext& ctx,
                               const std::vector<int32_t>& perm,
                               Tensor* out) {
   const int rank = perm.size();
-  auto ret = TransposeSimple<T>::run(ctx, in, perm, out);
+  int64_t numel = in.numel();
+  bool ret{false};
+  if (numel >= INT32_MAX) {
```
Comment (Xreki) [inline paddle/fluid/operators/transpose_op.cu.h:1186]: 建议使用`std::numeric_limits<int32_t>::max()`

**Review by Xreki — APPROVED** [review]
LGTM. 两个小建议。

**Pull request merged**